### ISSUE TITLE
ServerTimeoutsTest.testBlockingTimeoutWithSlowRead throws ReadPendingException [HTTPS]

### ISSUE THREAD

**Issue by sbordet** [eclipse/jetty.project#2424]
The test does not fail, but it throws `ReadPendingException`, which should not happen.
Some times, the test hangs forever.
DEBUG logs attached.
[2424.txt](https://github.com/eclipse/jetty.project/files/1891614/2424.txt)

**Comment by sbordet** [comment]
The problem stems from the fact that the `Connection` is read interested, but there is an async exception thrown (the `TimeoutException` thrown by `HttpInput.blockForContent()`).

This is similar to an idle timeout; however for an idle timeout we execute `AbstractEndPoint.onIdleExpired()` that has the logic to fail both the `FillInterest` and the `WriteFlusher` - failing the `FillInterest` would null out the read callback.

There is no direct access to the `FillInterest` in `AbstractEndPoint` from the `Connection` (we would need to downcast).

`HttpInput.blockForContent()` throwing the `TimeoutException` does fail the callback directly, but does not call `FillInterest.onFail()` that would null out the read callback.

The error handling mechanism kicks in, writes the `500` response, and because the `read()` has been woken up, we eventually return to `onFillable()` which would read zero (no content available, that is why it timed out in the first place), and tries to register for fill interest, causing the `ReadPendingException`.

I'm thinking we need a new API to handle async read exceptions that are not idle timeouts when we are fill interested.
This new API (in `EndPoint`) would need to at least call `FillInterest.onFail()` to null out the callback.

@gregw thoughts ?

**Comment by gregw** [comment]
@sbordet do you have any stack traces or a reproduction?  I'm finding it hard to follow exactly the sequence.

**Comment by gregw** [comment]
cancel that... I see the 2424.txt now...

**Comment by gregw** [comment]
@sbordet alternately, we could remove all the the timeout handling in HttpInput and just allow the normal endpoint timeout to occur and propagate up to the HttpInput.

**Comment by gregw** [comment]
... or just finally get rid of the blockingtimeout and just have the normal idle timeout

**Comment by gregw** [comment]
created a very speculative alternative thought bubble of a solution in https://github.com/eclipse/jetty.project/tree/jetty-9.4.x-2424-blockingTimeoutAbort

**Comment by sbordet** [comment]
@gregw I don't think it's ok to close the connection from `HttpInput.blockForContent()`.
The reason is that we want to be able to write a response 500 to the client.
Before your changes, the application would see an exception; either it handled it, or it is handled by Jetty, which kicks the error page mechanism.
After your changes, the connection is closed anyway, so no 500 arrives to the client (which is also the reason you had to change the test from `contains(" 500 ")` to `not(contains(" 200 "))` - the latter would match an empty string but that would not be ok).

**Comment by gregw** [comment]
replaced by #2525

**Comment by joakime** [comment]
Looks like the ReadPendingException is still present in the current `jetty-9.4.x` HEAD.

```
Running org.eclipse.jetty.server.ServerConnectorTimeoutTest.testHttpWriteIdleTimeout()
2018-05-16 06:38:17.258:INFO:oejs.Server:main: jetty-9.4.11-SNAPSHOT; built: 2018-05-16T11:27:23.267Z; git: 4cc8f4dcf6c64eade4d34b2456061a9b243f002b; jvm 9.0.4+11
2018-05-16 06:38:17.260:INFO:oejs.AbstractConnector:main: Started ServerConnector@684c2c57{HTTP/1.1,[http/1.1]}{0.0.0.0:55727}
2018-05-16 06:38:17.260:INFO:oejs.Server:main: Started @117820ms
2018-05-16 06:38:17.260:WARN:oejs.HttpConfiguration:main: HttpConfiguration.setBlockingTimeout is deprecated!
2018-05-16 06:38:17.764:WARN:oejs.HttpChannel:qtp2010094344-5533: /echo java.io.IOException: java.util.concurrent.TimeoutException: Blocking timeout 500 ms
2018-05-16 06:38:17.765:WARN:oeji.FillInterest:qtp2010094344-5533: Read pending for org.eclipse.jetty.server.HttpConnection$BlockingReadCallback@2f4c27b5 prevented AC.ReadCB@79d3e42b{HttpConnection@79d3e42b[p=HttpParser{s=CLOSE,10 of 20},g=HttpGenerator@1ea9abc2{s=START}]=>HttpChannelOverHttp@1dc36fb{r=1,c=false,a=IDLE,uri=null,age=0}<-SocketChannelEndPoint@51dabb30{/192.168.1.136:55728<->/192.168.1.136:55727,OSHUT,fill=FI,flush=-,to=1/2000}{io=1/1,kio=1,kro=1}->HttpConnection@79d3e42b[p=HttpParser{s=CLOSE,10 of 20},g=HttpGenerator@1ea9abc2{s=START}]=>HttpChannelOverHttp@1dc36fb{r=1,c=false,a=IDLE,uri=null,age=0}}
2018-05-16 06:38:17.766:WARN:oejut.QueuedThreadPool:qtp2010094344-5533:
java.nio.channels.ReadPendingException
	at org.eclipse.jetty.io.FillInterest.register(FillInterest.java:58)
	at org.eclipse.jetty.io.AbstractEndPoint.fillInterested(AbstractEndPoint.java:361)
	at org.eclipse.jetty.io.AbstractConnection.fillInterested(AbstractConnection.java:134)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:271)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:281)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:102)
	at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:118)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:760)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:678)
	at java.base/java.lang.Thread.run(Thread.java:844)
2018-05-16 06:38:17.766:WARN:oejut.QueuedThreadPool:qtp2010094344-5533: Unexpected thread death: org.eclipse.jetty.util.thread.QueuedThreadPool$2@29779f81 in QueuedThreadPool[qtp2010094344]@77cf9b08{STARTED,8<=8<=200,i=4,q=0}[ReservedThreadExecutor@7ad3c6d6{s=1/8,p=0}]
2018-05-16 06:38:18.269:INFO:oejs.AbstractConnector:main: Stopped ServerConnector@684c2c57{HTTP/1.1,[http/1.1]}{0.0.0.0:0}
```

**Comment by sbordet** [comment]
@lachlan-roberts the test is testing the blocking timeout, so can't be replaced with an idle timeout - although it does fix the `ReadPendingException`.
